### PR TITLE
Use base::SequenceBound for Rewards database

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/15779

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by szilardszaloki — APPROVED** [review]
👍🏻

File: ios/browser/api/ledger/brave_ledger.mm [hunk 1]
```diff
@@ -392,14 +372,13 @@ - (NSString*)rewardsDatabasePath {
 }
 
 - (void)resetRewardsDatabase {
-  delete rewardsDatabase;
```
Comment (bridiver) [inline ios/browser/api/ledger/brave_ledger.mm:395]: is this running on the main thread? If so this code was broken in terms of threading. If it's running on the task runner then it's not safe to initialize the new db here.

Comment (bridiver) [reply ios/browser/api/ledger/brave_ledger.mm:395]: if it's running on the main thread then the new code is ok

File: ios/browser/api/ledger/brave_ledger.mm [hunk 2]
```diff
@@ -244,10 +228,6 @@ - (instancetype)initWithStateStoragePath:(NSString*)path {
 - (void)dealloc {
   [NSNotificationCenter.defaultCenter removeObserver:self];
   [self.notificationStartupTimer invalidate];
-
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply ios/browser/api/ledger/brave_ledger.mm:247]: I was assuming that the destructor for `rewardsDatabase` would be automatically called, and so there would be no need to call 'Reset' explicitly. @kylehickinson is that right?

[the remark this answers is not in the document]
Comment (zenparsing) [reply ios/browser/api/ledger/brave_ledger.mm:247]: I was assuming that the destructor for `rewardsDatabase` would be automatically called, and so there would be no need to call 'Reset' explicitly. @kylehickinson is that right?

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -3167,29 +3162,14 @@ void RewardsServiceImpl::ReconcileStampReset() {
   }
 }
 
-ledger::type::DBCommandResponsePtr RunDBTransactionOnFileTaskRunner(
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/browser/rewards_service_impl.cc:3170]: Yes - `RunDBTransactionOnFileTaskRunner` has been removed here and in `brave_ledger.mm`.

[the remark this answers is not in the document]
Comment (bridiver) [reply components/brave_rewards/browser/rewards_service_impl.cc:3170]: sorry, I totally misread this diff :)

File: ios/browser/api/ledger/brave_ledger.mm (hunk 2 above)
[the remark this answers is not in the document]
Comment (bridiver) [reply ios/browser/api/ledger/brave_ledger.mm:247]: oh, you mean it gets called in the normal process of dealloc? That's probably correct and I guess really the problem here is that ledger and ledgerClient should be unique_ptr

File: ios/browser/api/ledger/brave_ledger.mm (hunk 1 above)
Comment (bridiver) [reply ios/browser/api/ledger/brave_ledger.mm:395]: @kylehickinson verified that this is run on the main thread and only in some debug mode so the new code is fine

**Review by bridiver — APPROVED** [review]

File: ios/browser/api/ledger/brave_ledger.mm (hunk 1 above)
Comment (kylehickinson) [reply ios/browser/api/ledger/brave_ledger.mm:395]: Yep always main thread, but also this code is only run in debug mode when resetting rewards (never run in production atm)

**Review by kylehickinson — APPROVED** [review]
iOS++

**Review by bsclifton — APPROVED** [review]
Approving as @szilardszaloki had already approved and I changed the `rewards-client` group permission (to assign group, not individuals) 😄

**Pull request merged**